Added new HLTfilter to select L3 and L2 muons

alexsr-98 opened the pull request:
#### PR description:

- This PR adds a HLTfilter (HLTMuonL3andL2PreFilter.cc) designed to filter L2 and L3 muons. If the L3 exists it is saved together with all the L2 muons which did not match with any L3. 
- There is the possibility to cut on different variables of the L2 and L3 muons independently.
- There are two different parameters to control the minimum number of L2 and L3 muons that you want to have in order to trigger the event. 
- This development has been presented in several Muon POG meetings: [presentation](https://indico.cern.ch/event/1096479/#3-status-of-global-displaced-m).

#### PR validation:

- Passed the "scram b runtests" test.

alexsr-98 committed:
Added new HLTfilter to select L3 and L2 muons

Applied patch for code quality checks

Apply suggestions from code review

Applied some suggestions from code review.

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Apply suggestions from code review

Applied more suggestions from code review

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Adressed first set of comments of the PR and added a customizer to implement the filter

Applied patch for code quality checks

Apply suggestions from code review

Applied some suggestions from code review

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Applied suggestions from code review

Applied code checks

Applied comments from the PR

Applied comment

Applied code checks

Apply suggestions from code review

Implemented some comments from PR review.

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Implemented comments from PR review.

Implemented code checks and removed customizer file

Implemented comments from PR review

Implemented code checks

Reorganized the code to make it clearer and removed repeated code

Applied code checks

Implemented comments from PR review

Applied code checks

Small fix in the fillDescriptions function

Applied comments from PR

Applied code checks

Renamed plugin

Applied code checks

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/26872

- This PR adds an extra 16KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/26872/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/26872/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

missirol commented:
@alexsr-98
This PR can't proceed without first passing code checks. Please follow
https://cms-sw.github.io/PRWorkflow.html

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/26895

- This PR adds an extra 20KB to repository

cmsbuild commented:
A new Pull Request was created by  @alexsr-98  for master.

It involves the following packages:

- HLTrigger/Muon (**hlt**)


@cmsbuild, @missirol, @Martin-Grunewald can you please review it and eventually sign? Thanks.
@Fedespring, @HuguesBrun, @calderona, @silviodonato, @abbiendi, @jhgoh, @Martin-Grunewald, @missirol, @sscruz, @CeliaFernandez, @trocino, @cericeci this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,343 @@
+    if (L2tk->normalizedChi2() > max_NormalizedChi2_L2_)
+      FailedL2 = true;
+
+    if (recoBeamSpotHandle.isValid()) {
```
```suggestion
    if (useBeamspot_) {
      if (not recoBeamSpotHandle.isValid()) { throw cms::Exception("") << ""; }
```
(the suggestion is pseudo-code..)

If the beamspot handle is invalid (e.g. due to a typo in the InputTag), certain cuts/selections are silently skipped. It would be safer to add a boolean `useBeamspot` in the PSet, and to throw an exception if `useBeamspot==true and recoBeamSpotHandle.isValid() == false`.

alexsr-98 replied:
I have added the exception but I think that it is not necessary to include another parameter to use the beam spot. If you want to use the cuts that are implemented under this bracket you can change for example the parameter max_Dz_. The beam spot is used only to compute Dz (the impact parameter in the z direction).

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,343 @@
+      const BeamSpot& beamSpot = *recoBeamSpotHandle;
+
+      //dz cut
+      if (std::abs((L2cand->vz() - beamSpot.z0()) -
+                   ((L2cand->vx() - beamSpot.x0()) * L2cand->px() + (L2cand->vy() - beamSpot.y0()) * L2cand->py()) /
+                       L2cand->pt() * L2cand->pz() / L2cand->pt()) > max_Dz_)
```
```suggestion
      if (std::abs((L2cand->vz() - beamSpot.z0()) -
                   ((L2cand->vx() - beamSpot.x0()) * L2cand->px() + (L2cand->vy() - beamSpot.y0()) * L2cand->py()) /
                       L2cand->pt() * L2cand->pz() / L2cand->pt()) > max_Dz_)
```
The expression is fairly hard to read. Can you add a comment with an explanation of this cut, and try to make the code more readable?

alexsr-98 replied:
Yes, it is a cut on the impact parameter of the muon in the z direction. I will make it clearer.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,343 @@
+        FailedL2 = true;
+
+      // dxy significance cut (safeguard against bizarre values)
+      if (min_DxySig_ > 0 &&
+          (L2tk->dxyError() <= 0 || std::abs(L2tk->dxy(beamSpot.position()) / L2tk->dxyError()) < min_DxySig_))
+        FailedL2 = true;
+
+      //dxy beamspot cut
+      float absDxy = std::abs(L2tk->dxy(beamSpot.position()));
```
```suggestion
      float const absDxy = std::abs(L2tk->dxy(beamSpot.position()));
      if (min_DxySig_ > 0 &&
          (L2tk->dxyError() <= 0 || (absDxy < min_DxySig_ * L2tk->dxyError())))
        FailedL2 = true;

      //dxy beamspot cut
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,343 @@
+      LogDebug("HLTMuonL3SimplePreFilter") << "cand isNonnull " << cand.isNonnull();
+
+      //did this candidate triggered at previous stage.
+      if (matchPreviousCand_ && !triggerdByPreviousLevel(cand, vcands))
```
```suggestion
      if (matchPreviousCand_ && !triggeredByPreviousLevel(cand, vcands))
```

missirol reviewed (commented):
A first round of comments. I have not completed yet the review of `HLTMuonL3orL2PreFilter::hltFilter`, but these can be addressed in the meantime.

Martin-Grunewald commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,343 @@
+  // this HLT filter, and place it in the Event.
+
+  if (saveTags())
+    filterproduct.addCollectionTag(candTag_);
```
I think you need to add also ```filterproduct.addCollectionTag(L2CandTag_);```
as you also want to record L2cands.

alexsr-98 replied:
Yes sorry, I will add it

Martin-Grunewald commented:
Sorry, but 'the code compiles' is insufficient as validation, please run a trigger path exercising this filter and check(!) that it performs as intended in terms of recording L2/L3 candidates and this is done (technical) correctly by adding to your HLT path with the new plugin also the HLTriggerFinalPath which does the packing up (and will complain with error messages in case of miscodings).

folguera commented:
@Martin-Grunewald we are in the process of developing a new reconstruction sequence for long-lived particles, and this filter has been fully validated in our development area with a mockup path that we are using for validation of efficiencies, rates and timing.The final HLT that we will propose will come after. 

The motivation for this PR is that we need to share the reconstruction sequence with other groups and we did plan to use confdb for this, and if I'm not mistaken, we would need this filter to be merged in order to be picked by confDB.

Martin-Grunewald commented:
OK, that's good. Still, more checking is needed as per my code comment above - you should have seen error messages arising from one of the modules in the HLTriggerFinalPath...

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/26941

ERROR: Build errors found during clang-tidy run.
```
         reco::BeamSpot
--
HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:115:40: error: use of undeclared identifier 'L2mucands'; did you mean 'mucands'? [clang-diagnostic-error]
  for (unsigned int iMuL2 = 0; iMuL2 < L2mucands->size(); iMuL2++) {
                                       ^~~~~~~~~
                                       mucands
--
HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:116:5: error: unknown type name 'RecoChargedCandidateRef'; did you mean 'reco::RecoChargedCandidateRef'? [clang-diagnostic-error]
    RecoChargedCandidateRef L2cand(L2mucands, iMuL2);
    ^~~~~~~~~~~~~~~~~~~~~~~
    reco::RecoChargedCandidateRef
--
HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:116:36: error: use of undeclared identifier 'L2mucands'; did you mean 'mucands'? [clang-diagnostic-error]
    RecoChargedCandidateRef L2cand(L2mucands, iMuL2);
                                   ^~~~~~~~~
                                   mucands
--
HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:126:5: error: unknown type name 'TrackRef'; did you mean 'reco::TrackRef'? [clang-diagnostic-error]
    TrackRef L2tk = L2cand->track();
    ^~~~~~~~
    reco::TrackRef
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/26942

ERROR: Build errors found during clang-tidy run.
```
HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:12:1: error: use of undeclared identifier 'HLTMuonL3orL2PreFilter' [clang-diagnostic-error]
HLTMuonL3orL2PreFilter::HLTMuonL3orL2PreFilter(const edm::ParameterSet& iConfig)
^
HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:46:6: error: use of undeclared identifier 'HLTMuonL3orL2PreFilter' [clang-diagnostic-error]
void HLTMuonL3orL2PreFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
     ^
HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:78:6: error: use of undeclared identifier 'HLTMuonL3orL2PreFilter' [clang-diagnostic-error]
bool HLTMuonL3orL2PreFilter::hltFilter(edm::Event& iEvent,
     ^
HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:299:6: error: use of undeclared identifier 'HLTMuonL3orL2PreFilter' [clang-diagnostic-error]
bool HLTMuonL3orL2PreFilter::triggeredByPreviousLevel(const reco::RecoChargedCandidateRef& candref,
     ^
HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:304:19: error: use of undeclared identifier 'HLTMuonL3orL2PreFilter' [clang-diagnostic-error]
DEFINE_FWK_MODULE(HLTMuonL3orL2PreFilter);
                  ^
Suppressed 1326 warnings (1325 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,343 @@
+ *
+ */
+
+#include "HLTMuonL3orL2PreFilter.h"
```
This comment has not been addressed (and should not have been marked as 'resolved'). Please merge the two files.

alexsr-98 commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,343 @@
+void HLTMuonL3orL2PreFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  makeHLTFilterDescription(desc);
+  desc.add<edm::InputTag>("BeamSpotTag", edm::InputTag("hltOfflineBeamSpot"));
```
It was a typo, It should be `hltOnlineBeamSpot`. It is now fixed.

missirol commented:
@alexsr-98

Since there was no update to this PR in the last two weeks, let me clarify: the review will continue only once a new version is committed addressing the first round of comments (and passing code-checks..).

alexsr-98 commented:
Sorry for the delay, we were working to produce a more complete validation of the filter by measuring: efficiencies, rates, timing and pT resolution. These results can be found here: [HLT-Muon-Displaced_2021-12-12.pdf](https://github.com/cms-sw/cmssw/files/7699265/HLT-Muon-Displaced_2021-12-12.pdf).

I have also added a new file `customizeHLTforDisplacedGlobal.py` to implement the filter.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27311

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27311/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27311/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27312

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

cmsbuild commented:
Pull request #36239 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,390 @@
+
+    ////////////////////////////////////////L3///////////////////////////////////////////////////////////
+    for (unsigned int iMu = 0; iMu < mucands.size(); iMu++) {
+      reco::RecoChargedCandidateRef cand(iEvent.getHandle(candToken_), iMu);
```
```suggestion
      reco::RecoChargedCandidateRef cand(iEvent.getHandle(candToken_), iMu);
```
The `getHandle` here is basically a repetition of what was used to initialise `mucands`. Since the handle is needed, you can use `auto const mucands = iEvent.getHandle(candToken_)` upstairs, and then replace `mucands.` with `mucands->` throughout, plus replace here `iEvent.getHandle(candToken_)` with `mucands`.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,390 @@
+      FailedL2 = true;
+
+    // Pt threshold cut
+    double L2pt = L2cand->pt();
+    double L2err0 = L2tk->error(0);
+    double L2abspar0 = std::abs(L2tk->parameter(0));
+    double L2ptLx = L2pt;
+    // convert 50% efficiency threshold to 90% efficiency threshold
+    if (L2abspar0 > 0)
+      L2ptLx += nsigma_Pt_ * L2err0 / L2abspar0 * L2pt;
```
Since this construct appears at least twice (see below), please include it in a function, to minimise repeated code.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,390 @@
+      }
+    }
+    if (failNstations || failNhits || failNchambers)
+      FailedL2 = true;
```
General question: when `FailedL2` is `true`, could you `continue` right away (and go to the next candidate), or do you really need to go through all the following steps in the loop scope?

alexsr-98 replied:
I think it is better to go through. For example, imagine an event with only one muon which is reconstructed in both ways, global and standalone. Because we may have different cuts in L2 and L3 muons, if this muon is rejected in the first loop of L2 muons, then, it will never enter in the second loop.

missirol reviewed (commented):
I still have to fully understand the logic of the filter, but in the meantime this is the second round of review comments.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27445

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27445/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27445/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

alexsr-98 commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,390 @@
+/** \class HLTMuonL3orL2PreFilter
```
I did this in the previous commit, I merged everything into `HLTMuonL3orL2PreFilter.cc`.

alexsr-98 commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,390 @@
+
+        // dxy significance cut (safeguard against bizarre values)
+        if (min_DxySig_ > 0 &&
+            (tk->dxyError() <= 0 || std::abs(tk->dxy(beamSpot.position()) / tk->dxyError()) < min_DxySig_))
```
I don't fully understand this change. In the original one, we are comparing the significance of the track with the minimum significance specified in the parameter set.

missirol replied:
The idea was just to have `A < B*C` instead of `A/B < C`, to avoid the edge case where `B` is zero and causes a crash in the 2nd implementation (there is no check that B is >0). My assumption is that "B" (i.e. `tk->dxyError()`) is always non-negative; if not, this suggestion would not apply anymore.

alexsr-98 commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,390 @@
+    // cut on number of stations
+    bool failNstations(false), failNhits(false), failNchambers(false);
+    for (unsigned int i = 0; i < nAbsetaBins; ++i) {
+      if (std::abs(L2cand->eta()) < absetaBins_L2_[i]) {
+        if (L2cand->track()->hitPattern().muonStationsWithAnyHits() < minNstations_L2_[i]) {
```
I don't understand why do we have to check this. What is the reason?

missirol replied:
The reason is fairly obvious. If the PSet is misconfigured, for example with `absetaBins_L2_` having a larger size than `minNstations_L2_`, then `minNstations_L2_[i]` in L237 will cause a segmentation violation. The same reasoning applies to the other relevant vectors. In general, the logic of this module demands these vectors have the same size, and that's why this needs to be checked explicitly in the ctor; in case those sizes differ, one can get a seg-fault (see above), or possibly wrong physics outputs in the worst case scenario.

missirol replied:
Just noting that this point has not been addressed yet, as far as I can see. It is a simple update, and one can follow the example linked above.

missirol replied:
@alexsr-98 

Since this point is not optional, I'm waiting for this to be addressed before proceeding further with the review.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27458

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27458/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27458/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27459

- This PR adds an extra 28KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

cmsbuild commented:
Pull request #36239 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,367 @@
+
+      // dxy significance cut (safeguard against bizarre values)
+      if (min_DxySig_ > 0 &&
+          (tk->dxyError() <= 0 || std::abs(tk->dxy(beamSpot.position()) / tk->dxyError()) < min_DxySig_))
```
Here, the same comment as L218 applies, i.e.
```
if (min_DxySig_ > 0 && (tk->dxyError() <= 0 || (std::abs(tk->dxy(beamSpot.position()) < min_DxySig_ * tk->dxyError())))
```

missirol commented on HLTrigger/Muon/python/customizeHLTforDisplacedGlobal.py:
```diff
@@ -0,0 +1,44 @@
+import FWCore.ParameterSet.Config as cms
+
+def customizeHLTforDisplacedGlobal(process):
+	'''
```
Please fix the indentation in this function, using 4 spaces for block, i.e.
```python
def customizeHLTforDisplacedGlobal(process):
    """comment
    """
    process.hltL3orL2fL1f0Filtered24Displaced = ...
        CandTag = cms.InputTag ...
        ...
    )
    ...
    return process
```

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27694

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

cmsbuild commented:
Pull request #36239 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27700

- This PR adds an extra 32KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27700/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27700/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
Pull request #36239 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on HLTrigger/Muon/python/customizeHLTforDisplacedGlobal.py:
```diff
@@ -0,0 +1,44 @@
+import FWCore.ParameterSet.Config as cms
+
+def customizeHLTforDisplacedGlobal(process):
+	'''
+	Temporal solution to adapt the HLT global displaced path with a new filter 
+	that selects dSA and dGbl (this customizer will not be neccesary when the 
+	filter is merged in the central repository).
+	'''
+	process.hltL3orL2fL1f0Filtered24Displaced = cms.EDFilter( "HLTMuonL3orL2PreFilter",
```
```suggestion
    '''
    Temporal solution to adapt the HLT global displaced path with a new filter 
    that selects dSA and dGbl (this customizer will not be neccesary when the 
    filter is merged in the central repository).
    '''
    process.hltL3orL2fL1f0Filtered24Displaced = cms.EDFilter( "HLTMuonL3orL2PreFilter",
```
See comment above on indentation.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,376 @@
+    return (deltaR2(r1, r2) < maxDeltaR2);
+  }
+
+  double pTeffTransformation(reco::RecoChargedCandidateRef candidate, reco::TrackRef track) const {
+    auto const pt = candidate->pt();
+    auto const err0 = track->error(0);
+    auto const abspar0 = std::abs(track->parameter(0));
+    // if abspar0 > 0, convert 50% efficiency threshold to 90% efficiency threshold
+    auto const ptLx = abspar0 > 0 ? pt + nsigma_Pt_ * err0 / abspar0 * pt : pt;
+    return ptLx;
+  }
```
```suggestion
  double pTeffTransformation(reco::RecoChargedCandidateRef const& candidate, reco::TrackRef const& track) const;
```
Please move implementation outside of class definition (see comment below).

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,376 @@
+                                                      const std::vector<reco::RecoChargedCandidateRef>& vcands) const {
+  return (std::find(vcands.begin(), vcands.end(), candref) != vcands.end());
+}
+
```
```suggestion

  double HLTMuonL3orL2PreFilter::pTeffTransformation(reco::RecoChargedCandidateRef const& candidate, reco::TrackRef const& track) const {
    auto const pt = candidate->pt();
    auto const err0 = track->error(0);
    auto const abspar0 = std::abs(track->parameter(0));
    // if abspar0 > 0, convert 50% efficiency threshold to 90% efficiency threshold
    auto const ptLx = abspar0 > 0 ? pt + nsigma_Pt_ * err0 / abspar0 * pt : pt;
    return ptLx;
  }

```

missirol reviewed (commented):
@alexsr-98 This PR is converging, but I have a few more comments on the latest version.

alexsr-98 commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,376 @@
+    auto const vz_L2 = L2cand->vz() - beamSpot.z0();
+    auto const vx_L2 = L2cand->vx() - beamSpot.x0();
+    auto const vy_L2 = L2cand->vy() - beamSpot.y0();
+    if (std::abs(vz_L2 - (vx_L2 * L2cand->px() + vy_L2 * L2cand->py()) / (L2cand->pt() * L2cand->pt()) * L2cand->pz()) >
```
Well, a muon with `pt` 0 will never reach the chambers and hence won't be reconstructed. I think this is not possible.

missirol replied:
My comment was just technical, not about physics. I would like to eliminate the chance that we silently get a NaN (0./0.) even in the case where wrong inputs are used (e.g. intentionally non-sense 4-vectors with pt=0), as unlikely as that might be. How to deal with this edge case (skip the candidate, throw an exception, or else) is left to you, proponent of this filter.

missirol replied:
Still to be addressed: safeguard against `pt=0`.

alexsr-98 replied:
Ok, I will skip it and add a message.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27785

- This PR adds an extra 32KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

cmsbuild commented:
Pull request #36239 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27787

- This PR adds an extra 32KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27787/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27787/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

Martin-Grunewald commented:
Sorry, but please have the added customisation routine/file removed from this PR, and instead, a JIRA ticket is filed to request the corresponding update.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/27790

- This PR adds an extra 36KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/python/customizeHLTforDisplacedGlobal.py:
    - Added: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
    - Modified: 57f38ce003c6fc87660fbdd5bf19a8c8bd356d1c, b329ff7c5d8b446e25e88b2ab7baaf765ad9737b
    - Deleted: 0d409174e6d1c73935d0cc2dc6eae0ab1349fc31
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

cmsbuild commented:
Pull request #36239 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,373 @@
+                 trigger::TriggerFilterObjectWithRefs& filterproduct) const override;
+
+private:
+  bool checkOverlap(const reco::RecoChargedCandidate r1,
+                    const reco::RecoChargedCandidate r2,
```
```suggestion
  bool checkOverlap(const reco::RecoChargedCandidate& r1,
                    const reco::RecoChargedCandidate& r2,
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,373 @@
+
+    bool matchedL3 = false;
+
+    ////////////////////////////////////////L3///////////////////////////////////////////////////////////
```
```suggestion
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,373 @@
+      if (std::abs(cand->eta()) > max_Eta_L3_)
+        continue;
+
+      reco::TrackRef tk = cand->track();
```
```suggestion
      reco::TrackRef const& tk = cand->track();
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,390 @@
+        n++;
+        filterproduct.addObject(trigger::TriggerMuon, cand);
+      }
+      ////CHECK HERE IF L2 AND L3 MUONS ARE THE SAME (now a deltaR matching is implemented between the L2 and L3 candidates, maybe this could be improved)
```
Answer to this question from earlier still missing.

missirol reviewed (commented):
Remaining comments on the latest version.

alexsr-98 commented:
The deltaR matching is working fine, I will remove that comment form the code.

cmsbuild commented:
Pull request #36239 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented:
I give one last comment on the current implementation, suggesting an optional change.

Based on its description ("filter to select L3 muons and L2 if they dont match with any L3"), I think this filter could be written in a much simpler/clearer way. In addition, there is boilerplate code which could be avoided: the selection on L3 and L2 candidates (cuts on pT, dxy, etc) is almost identical, and it could be moved to a common function (similarly to what was done during the review for the function `pTeffTransformation`). If I understood the logic of this filter correctly, a clearer implementation in pseudo-code could be the following:
```cpp
hltFilter(){

  // loop over L3 muons
  vector<int> l3_indices;
  for ( idxL3 : L3muons ){

    if (matchPreviousCand_ and !triggeredByPreviousLevel(cand, vcands))
      continue;

    bool passL3 = funcMuonPasses( [l3 thresholds] );

    if(passL3){
      n++
      addFilterObject( [save L3] );
      l3_indices.emplace_back(idxL3);
    }
  }

  // loop over L2 muons
  for ( idxL2 : L2muons ){

    bool passL2 = funcMuonPasses( [l2 thresholds] );

    bool overlapsWithL3 = false;
    for ( idxL3 : l3_indices ){

      if(L2overlapsWithL3( idxL2 , idxL3 )){
        overlapsWithL3 = true;
        break;
      }
    }

    if(passL2 and !overlapsWithL3){
      n++
      addFilterObject( [save L2] );
    }
  }

}
```

missirol commented:
@alexsr-98

Please provide feedback on the latest review comment (https://github.com/cms-sw/cmssw/pull/36239#issuecomment-1019187099), so that this PR can move forward.

alexsr-98 commented:
> @alexsr-98
> 
> Please provide feedback on the latest review comment ([#36239 (comment)](https://github.com/cms-sw/cmssw/pull/36239#issuecomment-1019187099)), so that this PR can move forward.

Sorry for the delay. I am implementing that comment and checking that everything is fine and nothing changes. I will update the PR today if everything is right.

missirol commented:
No problem. I just wanted to check that the comment was received.

alexsr-98 commented:
To implement the last comment I used a struct to organize the parameters that define the cuts for the L2 and L3 muons. I also added more parameters to have the same for both L2 and L3 and added the possibility to select different number of L2 and L3 muons: `MinNL2` and `MinNL3`.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28158

- This PR adds an extra 76KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/python/customizeHLTforDisplacedGlobal.py:
    - Added: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
    - Modified: 57f38ce003c6fc87660fbdd5bf19a8c8bd356d1c, b329ff7c5d8b446e25e88b2ab7baaf765ad9737b
    - Deleted: 0d409174e6d1c73935d0cc2dc6eae0ab1349fc31
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28158/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28158/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28162

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/python/customizeHLTforDisplacedGlobal.py:
    - Added: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
    - Modified: 57f38ce003c6fc87660fbdd5bf19a8c8bd356d1c, b329ff7c5d8b446e25e88b2ab7baaf765ad9737b
    - Deleted: 0d409174e6d1c73935d0cc2dc6eae0ab1349fc31
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

cmsbuild commented:
Pull request #36239 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,440 @@
+  desc.add<std::vector<int>>("MinNchambersL2", {1, 0});
+  desc.add<bool>("MatchToPreviousCand", true);
+  desc.add<double>("MaxDeltaR2L2L3", 0.0025);
+  descriptions.add("hltMuonL3orL2PreFilter", desc);
```
```suggestion
  edm::ParameterSetDescription descMuonSelL2;
  MuonSelectionCuts::fillPSetDescription(descMuonSelL2);
  descriptions.add("L2CandSelection", descMuonSelL2);

  edm::ParameterSetDescription descMuonSelL3;
  MuonSelectionCuts::fillPSetDescription(descMuonSelL3);
  descriptions.add("L3CandSelection", descMuonSelL3);

  descriptions.addWithDefaultLabel(desc);
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,440 @@
+  for (unsigned int iMuL3 = 0; iMuL3 < L3mucands->size(); iMuL3++) {
+    reco::RecoChargedCandidateRef L3cand(L3mucands, iMuL3);
+    LogDebug("HLTMuonL3orL2PreFilter") << "L3cand isNonnull " << L3cand.isNonnull();
+    // safeguard
+    if (L3cand->pt() == 0.0) {
+      LogDebug("HLTMuonL3orL2PreFilter") << "Skipping candidate: L3cand pT is 0";
+      continue;
+    }
```
```suggestion

```

alexsr-98 replied:
Do I have to remove this? In a previous comment you suggested to add this to avoid divisions by 0.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,440 @@
+    // did this candidate triggered at previous stage.
+    if (matchPreviousCand_ && !triggeredByPreviousLevel(L3cand, vcands))
+      continue;
+    // apply all the cuts
+    if (!applySelection(L3cand, beamSpot, cutParsL3))
+      continue;
```
```suggestion

    // apply all the cuts
    if (!applySelection(L3cand, beamSpot, cutParsL3))
      continue;

```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,440 @@
+  for (unsigned int iMuL2 = 0; iMuL2 < L2mucands->size(); iMuL2++) {
+    reco::RecoChargedCandidateRef L2cand(L2mucands, iMuL2);
+    LogDebug("HLTMuonL3orL2PreFilter") << "L2cand isNonnull " << L2cand.isNonnull();
+    // safeguard
+    if (L2cand->pt() == 0.0) {
+      LogDebug("HLTMuonL3orL2PreFilter") << "Skipping candidate: L2cand pT is 0";
+      continue;
+    }
```
```suggestion

```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,440 @@
+bool HLTMuonL3orL2PreFilter::checkOverlap(
+    const reco::RecoChargedCandidateRef& r1,
+    const std::vector<reco::RecoChargedCandidateRef>& R2,  // vector with the recoCharged candidates from L3
+    const double& maxDeltaR2) const {
+  for (unsigned int iMu = 0; iMu < R2.size(); iMu++) {
+    reco::RecoChargedCandidateRef r2 = R2[iMu];
```
```suggestion
    const double maxDeltaR2) const {
  for (auto const& r2 : R2) {
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,440 @@
+  return false;
+}
+
+double HLTMuonL3orL2PreFilter::pTeffTransformation(const reco::RecoChargedCandidateRef& candidate,
+                                                   const reco::TrackRef& track,
+                                                   const cutParameters& cutPars) const {
+  auto const pt = candidate->pt();
+  auto const err0 = track->error(0);
+  auto const abspar0 = std::abs(track->parameter(0));
+  // if abspar0 > 0, convert 50% efficiency threshold to 90% efficiency threshold
+  auto const ptLx = abspar0 > 0 ? pt + cutPars.nsigma_Pt * err0 / abspar0 * pt : pt;
+  return ptLx;
+}
```
```suggestion
```
In the latest version this is only called in one place, so it's easier to just move it inside `applySelection`.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,440 @@
+    return false;
+
+  // pt threshold cut
+  if (pTeffTransformation(candidate, tk, cutPars) < cutPars.min_Pt)
+    return false;
```
```suggestion
  auto const err0 = tk->error(0);
  auto const abspar0 = std::abs(tk->parameter(0));
  // if abspar0 > 0, convert 50% efficiency threshold to 90% efficiency threshold
  auto const ptLx = abspar0 > 0 ? candPt + cutPars.nsigma_Pt * err0 / abspar0 * candPt : candPt;
  if (ptLx < cutPars.min_Pt)
    return false;
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,440 @@
+  auto const vz = candidate->vz() - beamSpot.z0();
+  auto const vx = candidate->vx() - beamSpot.x0();
+  auto const vy = candidate->vy() - beamSpot.y0();
+  if (std::abs(vz - (vx * candidate->px() + vy * candidate->py()) / (candidate->pt() * candidate->pt()) *
+                        candidate->pz()) > cutPars.max_Dz)
```
```suggestion
  if (candidate->pt() <= 0 or (std::abs(vz - (vx * candidate->px() + vy * candidate->py()) / (candidate->pt() * candidate->pt()) *
                        candidate->pz()) > cutPars.max_Dz))
```

missirol replied:
After a 2nd look, I think it would be sufficient to add a condition like the one here to avoid the pT==0 case. If you prefer to keep the explicit check on "pT == 0" as in the previous implementation, that's also okay (but it should probably become a LogWarning instead of a LogDebug).

alexsr-98 replied:
Yes, I think the same. I didn't see this comment

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,440 @@
+  desc.add<double>("MaxDeltaR2L2L3", 0.0025);
+  descriptions.add("hltMuonL3orL2PreFilter", desc);
+}
+
```
```suggestion
void HLTMuonL3orL2PreFilter::MuonSelectionCuts::fillPSetDescription(edm::ParameterSetDescription& desc) {
  desc.add<int>("MinN", 1);
  desc.add<double>("MaxDXYBeamSpot", 9999.0);
  desc.add<double>("MinDXYBeamSpot", -1.0);
  desc.add<double>("MaxDz", 9999.0);
  desc.add<double>("MinDxySig", -1.0);
  desc.add<double>("NSigmaPt", 0.0);
  desc.add<double>("MaxPtDifference", 9999.0);
  desc.add<double>("MinTrackPt", 0.0);
  desc.add<double>("MaxNormalizedChi2", 9999.0);
  desc.add<double>("MinPt", 23.0);
  desc.add<double>("MaxEta", 2.5);
  desc.add<int>("MinNmuonHits", 0);
  desc.add<std::vector<int>>("MinNhits", {1, 0});
  desc.add<std::vector<double>>("AbsEtaBins", {1, 9999.});
  desc.add<std::vector<int>>("MinNstations", {1, 1});
  desc.add<std::vector<int>>("MinNchambers", {1, 0});
}
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,440 @@
+  cutParsL2.minNstations = minNstations_L2_;
+  cutParsL2.minNchambers = minNchambers_L2_;
+}
+
```
```suggestion
HLTMuonL3orL2PreFilter::MuonSelectionCuts(const edm::ParameterSet& iConfig)
      min_N(iConfig.getParameter<int>("MinN")),
      max_DXYBeamSpot(iConfig.getParameter<double>("MaxDXYBeamSpot")),
      min_DXYBeamSpot(iConfig.getParameter<double>("MinDXYBeamSpot")),
      max_Dz(iConfig.getParameter<double>("MaxDz")),
      min_DxySig(iConfig.getParameter<double>("MinDxySig")),
      nsigma_Pt(iConfig.getParameter<double>("NSigmaPt")),
      max_PtDifference(iConfig.getParameter<double>("MaxPtDifference")),
      min_TrackPt(iConfig.getParameter<double>("MinTrackPt")),
      max_NormalizedChi2(iConfig.getParameter<double>("MaxNormalizedChi2")),
      min_Pt(iConfig.getParameter<double>("MinPt")),
      max_Eta(iConfig.getParameter<double>("MaxEta")),
      min_NmuonHits(iConfig.getParameter<int>("MinNmuonHits")),
      min_Nhits(iConfig.getParameter<std::vector<int>>("MinNhits")),
      absetaBins(iConfig.getParameter<std::vector<double>>("AbsEtaBins")),
      minNstations(iConfig.getParameter<std::vector<int>>("MinNstations")),
      minNchambers(iConfig.getParameter<std::vector<int>>("MinNchambers")) {}

```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,440 @@
+
+#include <vector>
+
+struct cutParameters {  // parameters for both L2 and L3
+  double max_DXYBeamSpot;
+  double min_DXYBeamSpot;
+  double max_Dz;
+  double min_DxySig;
+  double nsigma_Pt;
+  double max_PtDifference;
+  double min_TrackPt;
+  double max_NormalizedChi2;
+  double min_Pt;
+  double max_Eta;
+  int min_NmuonHits;
+  std::vector<int> min_Nhits;
+  std::vector<double> absetaBins;
+  std::vector<int> minNstations;
+  std::vector<int> minNchambers;
+};
```
```suggestion
```
Moved inside the plugin.

missirol reviewed (commented):
Thanks for the new version, I think it looks clearer. The main suggestion is to reduce code duplication a bit further by using a PSet for the muon-selection cuts, and a corresponding helper class inside the plugin (similar to the struct you added in the latest version). I didn't test the code, so expect compilation errors (and code-format errors) at first if you simply apply the suggested changes via GitHub.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28182

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/python/customizeHLTforDisplacedGlobal.py:
    - Added: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
    - Modified: 57f38ce003c6fc87660fbdd5bf19a8c8bd356d1c, b329ff7c5d8b446e25e88b2ab7baaf765ad9737b
    - Deleted: 0d409174e6d1c73935d0cc2dc6eae0ab1349fc31
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28182/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28182/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28183

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/python/customizeHLTforDisplacedGlobal.py:
    - Added: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
    - Modified: 57f38ce003c6fc87660fbdd5bf19a8c8bd356d1c, b329ff7c5d8b446e25e88b2ab7baaf765ad9737b
    - Deleted: 0d409174e6d1c73935d0cc2dc6eae0ab1349fc31
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

cmsbuild commented:
Pull request #36239 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28184

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/python/customizeHLTforDisplacedGlobal.py:
    - Added: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
    - Modified: 57f38ce003c6fc87660fbdd5bf19a8c8bd356d1c, b329ff7c5d8b446e25e88b2ab7baaf765ad9737b
    - Deleted: 0d409174e6d1c73935d0cc2dc6eae0ab1349fc31
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

cmsbuild commented:
Pull request #36239 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+ *
+ */
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/HLTReco/interface/TriggerRefsCollections.h"
```
```suggestion
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+
+  bool applySelection(const reco::RecoChargedCandidateRef&, const reco::BeamSpot&, const MuonSelectionCuts&) const;
+
+  const edm::InputTag L3candTag_;                                             // input tag identifying L3 muon container
+  const edm::EDGetTokenT<reco::RecoChargedCandidateCollection> L3candToken_;  // token identifying L3 muon container
```
```suggestion
  const edm::InputTag L3CandTag_;                                             // input tag identifying L3 muon container
  const edm::EDGetTokenT<reco::RecoChargedCandidateCollection> L3CandToken_;  // token identifying L3 muon container
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+  const edm::EDGetTokenT<reco::BeamSpot> beamspotToken_;  // token identifying beamSpot container
+
+  const bool matchPreviousCand_;  // flag to match L2 or L3 candidates to the previous level
+  const double maxDeltaR2_L2L3_;  // maximum deltaR2 between L2 and L3 muon to consider them the same
```
```suggestion
  const double maxDeltaR_L2L3_;  // maximum deltaR between L2 and L3 muon to consider them the same
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+      beamspotTag_(iConfig.getParameter<edm::InputTag>("BeamSpotTag")),
+      beamspotToken_(consumes(beamspotTag_)),
+      matchPreviousCand_(iConfig.getParameter<bool>("MatchToPreviousCand")),
+      maxDeltaR2_L2L3_(iConfig.getParameter<double>("MaxDeltaR2L2L3")),
```
```suggestion
      maxDeltaR_L2L3_(iConfig.getParameter<double>("MaxDeltaRL2L3")),
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+        << "), \"MinNstations\" (" << cutParsL3_.minNstations.size() << "), \"MinNchambers\" ("
+        << cutParsL3_.minNchambers.size() << ") and \"MinNhits\" (" << cutParsL3_.min_Nhits.size() << ") differ";
+  }
+}
```
```suggestion

  if(maxDeltaR_L2L3_ <= 0.) {
    throw cms::Exception("Configuration") << "invalid value for parameter \"MaxDeltaRL2L3\" (must be > 0): " << maxDeltaR_L2L3_;
  }
}
```
It's more intuitive/convenient to use DeltaR as parameter in the configuration. The conversion to DeltaR2 can be done inside the plugin. Since a negative value does not make sense in this plugin (if i understand correctly), an exception can be thrown in that case in the ctor.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+  desc.add<edm::InputTag>("PreviousCandTag", edm::InputTag(""));
+  desc.add<edm::InputTag>("BeamSpotTag", edm::InputTag("hltOnlineBeamSpot"));
+  desc.add<bool>("MatchToPreviousCand", true);
+  desc.add<double>("MaxDeltaR2L2L3", 0.0025);
```
```suggestion
  desc.add<double>("MaxDeltaRL2L3", 0.05);
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+  filterproduct.getObjects(trigger::TriggerMuon, passingL3mucands);
+
+  // Number of objects passing the L2 Trigger:
+  int nL2 = 0;
+  for (unsigned int iMuL2 = 0; iMuL2 < L2mucands->size(); iMuL2++) {
```
```suggestion
  int nL2 = 0;
  auto const maxDeltaR2_L2L3 = maxDeltaR_L2L3_ * maxDeltaR_L2L3_;
  for (unsigned int iMuL2 = 0; iMuL2 < L2mucands->size(); iMuL2++) {
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+    if (matchPreviousCand_ && !triggeredByPreviousLevel(L2cand, vcands))
+      continue;
+    // check if L2 and L3 candidates are the same
+    if (checkOverlap(L2cand, passingL3mucands, maxDeltaR2_L2L3_)) {
```
```suggestion
    if (checkOverlap(L2cand, passingL3mucands, maxDeltaR2_L2L3)) {
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+    filterproduct.addCollectionTag(L2CandTag_);
+  }
+  // get hold of trks
+  auto const L3mucands = iEvent.getHandle(L3candToken_);
```
```suggestion
  auto const L3mucands = iEvent.getHandle(L3CandToken_);
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+  // recording any reconstructed physics objects satisfying (or not)
+  // this HLT filter, and place it in the Event.
+  if (saveTags()) {
+    filterproduct.addCollectionTag(L3candTag_);
```
```suggestion
    filterproduct.addCollectionTag(L3CandTag_);
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+                                            const reco::BeamSpot& beamSpot,
+                                            const MuonSelectionCuts& cutPars) const {
+  // eta cut
+  if (std::abs(candidate->eta()) > cutPars.max_Eta)
```
```suggestion
  auto const candAbsEta = std::abs(candidate->eta());
  if (candAbsEta > cutPars.max_Eta)
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+  // cut on number of stations
+  bool failNstations(false), failNhits(false), failNchambers(false);
+  for (unsigned int i = 0; i < nAbsetaBins; ++i) {
+    if (std::abs(candidate->eta()) < cutPars.absetaBins[i]) {
```
```suggestion
    if (candAbsEta < cutPars.absetaBins[i]) {
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+    return false;
+
+  // dxy significance cut (safeguard against bizarre values)
+  double const absDxy = std::abs(tk->dxy(beamSpot.position()));
```
```suggestion
  auto const absDxy = std::abs(tk->dxy(beamSpot.position()));
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+        failNhits = true;
+      }
+      if ((candidate->track()->hitPattern().dtStationsWithAnyHits() +
+               candidate->track()->hitPattern().cscStationsWithAnyHits() <
+           cutPars.minNchambers[i])) {
```
```suggestion
               candidate->track()->hitPattern().cscStationsWithAnyHits()) <
           cutPars.minNchambers[i]) {
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+  }
+
+  // filter decision
+  const bool acceptL3(nL3 >= cutParsL3_.min_N);
+  const bool acceptL2(nL2 >= cutParsL2_.min_N);
```
If I remember correctly, the final filter result was based on the sum of L3 and L2 (L2+L3 > min_N). I thought this was the reason to name the plugin "L3orL2". Now this changes to two requirements (L2 > minN_L2 and L3 > minN_L3). I just want to ask if this is intended, and if the name of the plugin should change based on this.

alexsr-98 replied:
Yes I think we should change the name, but in the past it wasn't simply L2+L3, it saved all the L3 and all the L2 that didn't match to any L3. Now it does the same but also controlling the number of L2 and L3 independently.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,349 @@
+    reco::RecoChargedCandidateRef L2cand(L2mucands, iMuL2);
+    LogDebug("HLTMuonL3orL2PreFilter") << "L2cand isNonnull " << L2cand.isNonnull();
+    // did this candidate triggered at previous stage.
+    if (matchPreviousCand_ && !triggeredByPreviousLevel(L2cand, vcands))
+      continue;
```
If I remember well, in the original implementation the "triggeredByPreviousLevel" requirement was applied only on the L3 muons. Does it make sense to apply it to both L2Cands and L3Cands comparing against the same collection (i.e. "previousLevelCands")?

missirol reviewed (commented):
A couple of questions on the logic of the plugin, and some minor final technical suggestions.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28191

- This PR adds an extra 88KB to repository

- Found files with invalid states:
  - HLTrigger/Configuration/python/customizeHLTforDisplacedGlobal.py:
    - Added: 3b8377b02768d553f8d78414920f5acedc4f979d
    - Deleted: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
  - HLTrigger/Muon/python/customizeHLTforDisplacedGlobal.py:
    - Added: d8a59e46cdbb39709c7cbb3ea8512ac847abcda6
    - Modified: 57f38ce003c6fc87660fbdd5bf19a8c8bd356d1c, b329ff7c5d8b446e25e88b2ab7baaf765ad9737b
    - Deleted: 0d409174e6d1c73935d0cc2dc6eae0ab1349fc31
  - HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.h:
    - Added: 0e21f9703665632bcbe569ec60f7a9fdb20b7980
    - Modified: c7fb2d3bf08771be8592f6e4f45332af73b266bc, c279d6a72107b47789207dad7740086e7a6355f2
    - Deleted: 3b8377b02768d553f8d78414920f5acedc4f979d

cmsbuild commented:
Pull request #36239 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc:
```diff
@@ -0,0 +1,354 @@
+  // get hold of trks
+  auto const L3mucands = iEvent.getHandle(L3CandToken_);
+  auto const L2mucands = iEvent.getHandle(L2CandToken_);
+  auto const previousLevelCands = iEvent.getHandle(previousCandToken_);
+  std::vector<reco::RecoChargedCandidateRef> vcands;
+  if (previousLevelCands.isValid()) {
+    previousLevelCands->getObjects(trigger::TriggerMuon, vcands);
+  }
```
```suggestion
  std::vector<reco::RecoChargedCandidateRef> vcands;
  if (matchPreviousCand_) {
    auto const& previousLevelCands = iEvent.get(previousCandToken_);
    previousLevelCands.getObjects(trigger::TriggerMuon, vcands);
  }
```
I noticed that the vector needs to be filled only if `matchPreviousCand_ == true`. And when that is the case, the collection should be always valid; if not, the fwk should return an error (this is what the suggested update does, the validity check is done inside `iEvent.get`).

missirol reviewed (commented):
@alexsr-98 

I have one last technical comment.

Based on [your earlier comment](https://github.com/cms-sw/cmssw/pull/36239/files#r800476175), in your next update please also rename the plugin appropriately.

Lastly, before I approve this PR, please update this branch squashing all the commits of this PR into a single commit (so, making a single commit with the final version and force-pushing it to the branch associated to this PR; if you need help with this, feel free to ask).

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28212

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28212/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28212/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36239/28213

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #36239 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented:
please test

@alexsr-98 , the implementation looks okay to me, thanks.

The PR tests here won't exercise this new plugin (other than compiling it), so I assume you verified in your HLT studies that the latest version of the plugin produces the intended physics results (if not, please do verify this).

Also, since the plugin content has slightly changed during the review ("or" -> "and"), please update the PR title and description accordingly if need be.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-75a98a/22282/summary.html
**COMMIT**: 00be3e98860de79be317a9e439feb4966b913159
**CMSSW**: CMSSW_12_3_X_2022-02-07-2300/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36239/22282/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3765080
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3765050
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 45 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 193 log files, 42 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2022-02-07-2300+75a98a/48271/triggerResults): no differences found

alexsr-98 commented:
> please test
> 
> @alexsr-98 , the implementation looks okay to me, thanks.
> 
> The PR tests here won't exercise this new plugin (other than compiling it), so I assume you verified in your HLT studies that the latest version of the plugin produces the intended physics results (if not, please do verify this).
> 
> Also, since the plugin content has slightly changed during the review ("or" -> "and"), please update the PR title and description accordingly if need be.

Yes, I've checking it during the revision and it behaves as expected.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
